feat(lint): useTsExpectError for TS

antonk52 opened the pull request:
## Summary

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

Added a new linting rule that bans usage of `@ts-ignore` suppression comment in favour of `@ts-expect-error`. Since ts-ignore does not notify the user when the error goes away there's a chance that a new error can crawl up on the same line and the user won't know about it. After implementing this rule and applying an auto fix five `@ts-ignore`'s were found to be redundant which proves rule's usefulness.

<!-- Link any relevant issues if necessary or include a transcript of any Discord discussion. -->

In [`eslint-plugin-typescript`](https://www.npmjs.com/package/@typescript-eslint/eslint-plugin) there's a rule called `ban-ts-comment` with banned `ts-ignore` by [default](https://github.com/typescript-eslint/typescript-eslint/blob/2b75c11d69bee88ca0cb77d7efd32b8d0387e6b3/packages/eslint-plugin/src/rules/ban-ts-comment.ts#L91-L94).

## Test Plan

Test cases are included in this PR

antonk52 commented:
A warning: there're "flaky" tests that get cancelled (please see the logs for GH actions in [Rome ci ubuntu](https://github.com/rome/tools/pull/1473/checks?check_run_id=2540410615#step:4:18) and [Rome test macos](https://github.com/rome/tools/pull/1473/checks?check_run_id=2540410687#step:4:23)). 

Multiple linting rules' tests get canceled due to not handling `UIDPath` in [internal/binary-transport/RSERWriterBase.ts](https://github.com/rome/tools/blob/main/internal/binary-transport/RSERWriterBase.ts#L483). Not sure why the actions display the green check as if they've passed.

ematipico commented:
Could you provide a case where the expect error is preferred to the ignore? Code wise. I think I didn't understand the difference.

Also, how it's decided when to use the ignore and when to use the expect error?

For example, the top level await we have in some code is not a real error. Basically TS doesn't support the syntax. So we don't expect to have an error.

On a side note, we should change the rule name. The rules that have "prefer" are old rules and we didn't migrate them yet. We should use "no" or "use" as prefixes. Probably we should write it down in the contribution file.

ematipico commented:
> A warning: there're "flaky" tests that get cancelled (please see the logs for GH actions in [Rome ci ubuntu](https://github.com/rome/tools/pull/1473/checks?check_run_id=2540410615#step:4:18) and [Rome test macos](https://github.com/rome/tools/pull/1473/checks?check_run_id=2540410687#step:4:23)). 
> 
> Multiple linting rules' tests get canceled due to not handling `UIDPath` in [internal/binary-transport/RSERWriterBase.ts](https://github.com/rome/tools/blob/main/internal/binary-transport/RSERWriterBase.ts#L483). Not sure why the actions display the green check as if they've passed.

When you have an error like this, it means that the tests cases you have inside a rule have invalid syntax. 

I also don't know why the CI is green.

antonk52 commented:
Thank you for your review @ematipico 

Here're a couple of examples where `ts-expect-error` is preferred

1. Unexpected error

First we have an actual error. We know that TS would mark this as an error so we add a suppression
```ts
// @ts-ignore
let foo: boolean = 'string'
```

Some time passes and someone changes the code, so the error goes away
```ts
// @ts-ignore
let foo: boolean = true
```

Note that unlike [flow](https://flowtype.org) typescript does not mark any suppression as unused. That's partially why `ts-expect-error` was introduced.

More time passes by and a new error is introduced, but it wasn't reported since we had `ts-ignore`suppression above which does not care whether there is or not an error on the line below it
```ts
// @ts-ignore
let foo: boolean = null
```

This example used an oversimplified code to make a point. The suppression can be placed above a longer line where multiple errors can be suppressed without developer's knowledge. So the main point of preferring to use `ts-expect-error` is to be notified when the error is no longer there.

2. Unsupported syntax / runtime

The example with top level await is really good here. On some point typescript will start to support this feature and once Rome's codebase is updated it would be nice to also be notified of no longer needed suppressions.

You can also take a look at [the reasons](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#ts-ignore-or-ts-expect-error) listed by typescript team which suppression to pick.

antonk52 commented:
@ematipico I've addressed your comments and resolved conflicts with current main.

Also added a line to `CONTRIBUTING.md` about lint rules starting with `no` or `use`.

I haven't got to figure out why the tests were failing for my rule. The output is always the same ([gist](https://gist.github.com/antonk52/ae8b38bf1b8bdb02f65c5f1c7652df60)). I've tried having it as `*.test.rjson` and `*.test.ts` but the result is the same. To keep the CI green there is currently a `useTsExpectError.test.ts` with no actual cases in it 😔. I found a similar thing with `useInterfaces`. Not a fan of merging this with no tests but wasn't able to figure out the root cause of the error. Might need a hint or two for that.

Besides the tests it's looking good on my end.

ematipico commented:
Found the issue (we need a workaround as I suspected). At the moment when we test each category, we don't actually isolate the test to run a single rule, but the worker runs a full fledged linter and it applies all the rules to each test case. While doing so, another rule goes in conflict with the rule you try to test.

I noticed this issue while I was changing rules some time ago. I will be able to fix it, but now. Maybe when I will have more time.

In order to fix your test cases, you could do:

```diff
filename: "filename.ts"
invalid: [
	"
	// @ts-ignore
	let foo: boolean = 1;
++	foo;
	"
	"
	// @ts-ignore
	let foo: boolean = 1;
++	foo;
	"
]
valid: [
	"
	// @ts-ignore
	let foo: boolean = 1;
++	foo;
	"
]
```

antonk52 commented:
@ematipico I'm sorry, this is quite an unlucky rule. I've added `foo` and being faced with a new kind of error. From the [message](https://github.com/rome/tools/pull/1473/checks?check_run_id=2595636361#step:4:22) it seems like the rule is not being applied in the test. However, if I add the `// @ts-ignore` comment to an existing or a new file in Rome codebase and run `./rome check` against it, it will complain about the suppression. Not sure about the `internal/process/task_queues.js` file mentioned in the CI output. I might be able to check further some time next week.

ematipico commented:
I just merged a PR that fixes the test of lint rules.

https://github.com/rome/tools/pull/1504

You can now remove the workaround from your test cases.

antonk52 commented:
@ematipico

This rule is seem to be done now, your fix really did help 🙌

Side note: there was one error that I came across that even though the visitor worked as expected the test requires the diagnostic to be attached. I was thinking if perhaps we can have `createLintVisitor` instead of `createVisitor` that would only allow us to return `signals.remain` or `signals.replace` with diagnostics. This would help to avoid such problems in the future and hint people working on linting rules before the tests run since typescript should be able to catch such mistakes. This can be done in a separate PR. Interested to hear your thoughts.

antonk52 commented:
Just a friendly ping. Is there something blocking this PR or does it need more triage? 🙂

ematipico commented:
> Just a friendly ping. Is there something blocking this PR or does it need more triage? 🙂

I will take a look at it next week. About your discussion around the visitors, I didn't quite catch what you mean. Feel free to open an issue!

ematipico commented on website/src/docs/lint/rules/ts/useTsExpectError.md:
```diff
@@ -0,0 +1,187 @@
+title: Lint Rule ts/useTsExpectError
+layout: layouts/rule.liquid
+showHero: false
+description: MISSING DOCUMENTATION
```
Would it be possible to add a tiny description here?

ematipico commented:
Could you change title of the PR too? It should `useExpectError`

yassere commented:
Interestingly, there's an uncaught (unnecessary) `@ts-ignore` in `internal/resources/Resource.ts` because `resources` is included in the `lint.ignore` config so that code never gets checked, and it has multiple lint errors.

Thanks for helping reveal that problem!

antonk52 commented:
Thanks everyone for the review 🙂

I will apply the changes on the weekend once I get some time to work on it

antonk52 commented on internal/compiler/lint/rules/ts/useTsExpectError.ts:
```diff
@@ -0,0 +1,51 @@
+};
+
+export default createVisitor({
+	name: "ts/useTsExpectError",
```
- I used `ts` in the name since it's a part of the suppression comment.
- I don't have a strong opinion here but perhaps `suppression` can be a better word to use here since we are dealing not with an arbitrary comment but a suppression annotation. Maybe this can ease the understanding of what the rule is doing, ie `useExpectErrorSuppression` so we would not need `ts` in the name and the meaning would be more apparent

Either way happy to settle with any solution here

antonk52 commented:
Merged in the fresh main branch and addressed the comments.

One of the tests uses `/** @ts-ignore*/` and the formatter(not my rule) removes the extra leading * in the beginning, which makes it rather a potentially useless test case since the one above checks for a single line block comment suppression. Should I remove it?

PS: Thanks for the review!

ematipico commented:
> Merged in the fresh main branch and addressed the comments.
> 
> One of the tests uses `/** @ts-ignore*/` and the formatter(not my rule) removes the extra leading * in the beginning, which makes it rather a potentially useless test case since the one above checks for a single line block comment suppression. Should I remove it?
> 
> PS: Thanks for the review!

We can keep it. The change will be automatic once we will fix the formatter.
